File: workspace/agent/internal/sessionx/session_files.go
package sessionx

// Aggregation of "the files this session changed" (docs/log/68 / decisions/0049).
//
// The population is the edit tool calls in the transcript; the git working-tree state is
// joined in by the Console against GET /fs/changes, keyed by `(repo, rel)`. Only the former
// is produced here.
//
// It rides along where the todos do (CollectTasks → resp["tasks"]): the whole transcript, not
// a window, is folded and included in the same /messages response. No dedicated endpoint and
// no second poll (decision 3).

import (
	"net/http"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/k-k1/agent-fleet/workspace/agent/internal/gitx"
	"github.com/k-k1/agent-fleet/workspace/agent/internal/httpx"
	"github.com/k-k1/agent-fleet/workspace/agent/internal/session"
	"github.com/k-k1/agent-fleet/workspace/agent/internal/transcript"
)

// fileAggEntry is one session's folded state. Transcripts are append-only, so the answer
// for a prefix never changes and only the new tail has to be folded on each poll — which
// is what makes running a line differ over the whole history affordable at poll rate.
type fileAggEntry struct {
	src   string // transcript path; a different file means a different conversation
	head  string // fingerprint of the first record — catches a rewrite in place
	done  int    // how many lines/turns are already folded into acc
	acc   map[string]*transcript.FileTouch
	order []string // insertion order of acc's keys (stable tiebreak)
	used  time.Time
}

var (
	fileAggMu sync.Mutex
	fileAggs  = map[string]*fileAggEntry{}
)

// maxTrackedFiles bounds one session's list. A session that genuinely touches more than
// this is a mass rewrite, where a file list is not the affordance anyone wants; the cap
// keeps the response and the cache bounded rather than trying to serve that case.
const maxTrackedFiles = 500

// sessionFileTouches folds fresh transcript records into the session's accumulator and
// returns the current list, newest touch first.
//
// `stable` is how many of the records are IMMUTABLE. claude's jsonl lines never change
// once written, so it passes len(lines). The store-backed agents can still append parts
// to their last message (opencode does — see genericMutableTail), so they hold that one
// back: everything below `stable` is folded into the cache, and the mutable tail is
// folded into a copy on every call.
func sessionFileTouches(name, src, head string, n, stable int, edits func(from, to int) []transcript.FileEdit) []transcript.FileTouch {
	if stable < 0 {
		stable = 0
	}
	if stable > n {
		stable = n
	}
	fileAggMu.Lock()
	defer fileAggMu.Unlock()

	e := fileAggs[name]
	// A different transcript file, a rewritten head, or a transcript that SHRANK (reset,
	// replaced conversation, compaction into a new file) invalidates the fold — start over.
	if e == nil || e.src != src || e.head != head || e.done > stable {
		e = &fileAggEntry{src: src, head: head, acc: map[string]*transcript.FileTouch{}}
		fileAggs[name] = e
	}
	if e.done < stable {
		foldFileEdits(e, edits(e.done, stable))
		e.done = stable
	}
	e.used = time.Now()
	evictFileAggs()

	out := e
	if stable < n {
		out = cloneFileAgg(e)
		foldFileEdits(out, edits(stable, n))
	}
	list := make([]transcript.FileTouch, 0, len(out.order))
	for _, k := range out.order {
		if t := out.acc[k]; t != nil {
			list = append(list, *t)
		}
	}
	// Newest touch first: that is the question the list actually answers ("the one I just
	// changed"). Ties fall back to the transcript index, then insertion order (sort.Stable).
	sort.SliceStable(list, func(i, j int) bool {
		if list[i].LastTS != list[j].LastTS {
			return list[i].LastTS > list[j].LastTS
		}
		return list[i].LastIdx > list[j].LastIdx
	})
	return list
}

// foldFileEdits merges raw edit calls into per-file rows.
func foldFileEdits(e *fileAggEntry, edits []transcript.FileEdit) {
	for _, ed := range edits {
		abs := absEditPath(ed.Path, ed.Cwd)
		if abs == "" {
			continue
		}
		t := e.acc[abs]
		if t == nil {
			if len(e.acc) >= maxTrackedFiles {
				continue
			}
			repo, rel := repoRelOf(abs)
			t = &transcript.FileTouch{
				Path: toBrowseRel(abs, "", browseRoot()), Repo: repo, Rel: rel,
				Sidechain: ed.Sidechain,
			}
			e.acc[abs] = t
			e.order = append(e.order, abs)
		}
		t.Count++
		t.Added += ed.Added
		t.Removed += ed.Removed
		t.Verb = ed.Verb // the last call is what the file IS now
		// Sidechain marks a file only SUBAGENTS touched; one main-thread edit clears it.
		t.Sidechain = t.Sidechain && ed.Sidechain
		if ed.TS > t.LastTS || (ed.TS == t.LastTS && ed.Idx >= t.LastIdx) {
			t.LastTS, t.LastIdx = ed.TS, ed.Idx
		}
	}
}

func cloneFileAgg(e *fileAggEntry) *fileAggEntry {
	c := &fileAggEntry{src: e.src, head: e.head, done: e.done,
		acc: make(map[string]*transcript.FileTouch, len(e.acc)), order: append([]string(nil), e.order...)}
	for k, v := range e.acc {
		cp := *v
		c.acc[k] = &cp
	}
	return c
}

// absEditPath anchors the target the agent wrote. A relative path needs the turn's cwd;
// without one there is no way to place it, and a guess would open the wrong file in
// another working copy — so it is dropped instead.
func absEditPath(p, cwd string) string {
	p = strings.TrimSpace(p)
	if p == "" {
		return ""
	}
	if !filepath.IsAbs(p) {
		if cwd == "" {
			return ""
		}
		p = filepath.Join(cwd, p)
	}
	return filepath.Clean(p)
}

// repoRelOf splits an absolute path into the working-copy folder and the path inside it.
// Both are empty for anything outside ~/repos (a file in the home dir, an agent config):
// the row is still listed, it just has no git side to be joined with.
func repoRelOf(abs string) (repo, rel string) {
	r, err := filepath.Rel(gitx.ReposRoot(), abs)
	if err != nil || r == "." || r == ".." || strings.HasPrefix(r, ".."+string(filepath.Separator)) {
		return "", ""
	}
	parts := strings.SplitN(filepath.ToSlash(r), "/", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return "", ""
	}
	return parts[0], parts[1]
}

// evictFileAggs keeps the cache bounded. Entries are small (a few hundred rows), so this
// only has to stop unbounded growth across a long-lived Agent that has seen many sessions.
func evictFileAggs() {
	const keep = 64
	if len(fileAggs) <= keep {
		return
	}
	cut := time.Now().Add(-30 * time.Minute)
	for k, v := range fileAggs {
		if v.used.Before(cut) {
			delete(fileAggs, k)
		}
	}
	for len(fileAggs) > keep {
		oldest, at := "", time.Time{}
		for k, v := range fileAggs {
			if oldest == "" || v.used.Before(at) {
				oldest, at = k, v.used
			}
		}
		delete(fileAggs, oldest)
	}
}

// forgetSessionFiles drops a session's folded state (stop / delete / recreate). Losing it
// is never wrong — the next poll rebuilds from the transcript.
func forgetSessionFiles(name string) {
	fileAggMu.Lock()
	delete(fileAggs, name)
	fileAggMu.Unlock()
}

// fileAggHead fingerprints the first record so a transcript rewritten IN PLACE (same
// path, same or greater length) is not mistaken for an append.
func fileAggHead(b []byte) string {
	if len(b) > 256 {
		b = b[:256]
	}
	return string(b)
}

// ── Deciding what is committed (docs/log/68 P2) ──────────────────────────────────
//
// A row in the strip means "the agent edited this", yet some rows have no diff left in the
// working tree. P0 showed all of those as the single word "no changes", because committed and
// reverted could not be told apart.
//
// What is added here is only "committed"; "reverted" is never claimed. There are other reasons
// for having no diff and not appearing in a commit (it was already in a commit from before the
// session started, it happened in another working copy, the file was moved). The only fact
// that can be asserted is "it appeared in a commit"; calling the rest "reverted" would write an
// unfounded claim into the UI. Those rows stay "no changes".

// maxCommitScan bounds the log walk. A session that produced more commits than this is
// well past the point where a per-file badge is the interesting information.
const maxCommitScan = 200

// HandleSessionCommittedFiles reports the repo-relative paths that appeared in a commit
// in this session's working copy SINCE the session was created.
//
// The evidence is a timestamp, so commits made in parallel by another session in the same
// working copy are included too. That does little harm because the result is only joined
// against the files THIS session edited, and "a file you touched was committed afterwards" is
// true regardless of who committed it.
func HandleSessionCommittedFiles(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	if !session.ValidName(name) {
		httpx.WriteErr(w, http.StatusBadRequest, "bad_name", "invalid session name")
		return
	}
	meta, ok := session.ReadMeta(name)
	if !ok {
		httpx.WriteErr(w, http.StatusNotFound, "not_found", "no such session: "+name)
		return
	}
	httpx.WriteJSON(w, http.StatusOK, map[string]any{
		"repo":      meta.Repo,
		"committed": committedSince(meta.Dir, meta.CreatedAt),
	})
}

// committedSince lists the paths touched by commits in dir since `since` (RFC3339).
// Any failure — not a git working copy, no such directory, an unparsable timestamp —
// returns an empty list: the badge simply stays "no changes", which is the honest degradation.
func committedSince(dir, since string) []string {
	if dir == "" || since == "" {
		return []string{}
	}
	if _, err := time.Parse(time.RFC3339, since); err != nil {
		return []string{}
	}
	out, err := gitx.Run(dir, "-c", "core.quotePath=false", "log",
		"--since="+since, "--max-count="+strconv.Itoa(maxCommitScan),
		"--name-only", "--pretty=format:", "--no-renames")
	if err != nil {
		return []string{}
	}
	seen := map[string]bool{}
	paths := []string{}
	for _, ln := range strings.Split(out, "\n") {
		p := strings.TrimSpace(ln)
		if p == "" || seen[p] {
			continue
		}
		seen[p] = true
		paths = append(paths, p)
	}
	return paths
}

File: console/src/features/mirror/sessionFiles.ts
// sessionFiles — the model behind "Changed files" (docs/log/68): the files THIS session's
// agent edited, joined with the working tree's current git state.
//
// Two sources, deliberately:
//   - the population comes from the transcript (`files` on GET /sessions/{name}/messages),
//     which is the only thing that knows the session axis. It is aggregated over the
//     WHOLE transcript server-side — the turns the mirror holds are a tail window, so
//     anything counted here would undercount and then grow as the reader scrolls up.
//   - each row's STATE comes from GET /fs/changes, because the transcript only records
//     that an edit happened, never that it was later reverted or committed.
//
// The join key is (repo, rel), never `path`: browse-relative paths are rooted at the
// Agent's browse root (AF_BROWSE_ROOT) while /fs/changes always reports
// "repos/<repo>/<rel>". They agree by default, which is exactly why a mismatch would go
// unnoticed. See decisions/0049 decision 2.
import { create } from "zustand";
import { compareText } from "../../lib/intl.ts";
import type { MsgKey } from "../../lib/i18n/index.ts";
import { openFileDiff } from "../scm/open.ts";
import { openFileMode } from "../viewer/openFile.ts";
import { useLayoutStore } from "../../layout/store.ts";

/** One file the session edited, as the Agent aggregates it (transcript.FileTouch). */
export interface SessionFile {
  path: string; // browse-root relative — what the file view opens
  repo?: string; // working-copy folder; absent = outside ~/repos
  rel?: string; // repo-relative; absent = outside ~/repos
  verb: string; // "edit" | "add" | "delete"
  added?: number;
  removed?: number;
  count: number; // how many edit calls touched it
  lastIdx: number;
  lastTs?: string;
  sidechain?: boolean; // ONLY subagents touched it
}

/** One working-tree change from GET /fs/changes. */
export interface FsChange {
  path: string;
  repo: string;
  untracked?: boolean;
  index?: string;
  worktree?: string;
}

/**
 * What the working tree says about a file the session edited.
 *   unstaged/staged/untracked — git has a diff to show
 *   committed — no working-tree diff left, but the path appeared in a commit made since
 *             this session started (docs/log/68 P2).
 *   clean   — no working-tree diff and no such commit. This is NOT "reverted": the
 *             edit may have landed in an older commit, in another working copy, or under
 *             a name git no longer reports. Only the positive claim is made; the rest
 *             stays "No diff". And these rows are KEPT either way — dropping them reads
 *             to the user as "I just edited that and it is not in the list", i.e. as a
 *             broken feature.
 *   outside — edited outside ~/repos (an agent config, a file in the home dir). Listed,
 *             but there is no git side and no diff to open.
 */
export type FileState = "unstaged" | "staged" | "untracked" | "committed" | "clean" | "outside";

export interface FileRow extends SessionFile {
  name: string; // basename, the row's main label
  dir: string; // repo-relative directory, shown faintly after the name ("" = repo root)
  state: FileState;
  deleted: boolean; // nothing left to open (git says D, or the agent deleted it)
}

export type FileSort = "recent" | "path";

const changeKey = (repo: string, rel: string) => repo + "\u0000" + rel;

/** The porcelain column that decides a change's state (worktree wins, then index) — the
 *  same rule the rail's Changes list uses, so the two never disagree about a badge. */
const codeOf = (c: FsChange): string => (c.worktree && c.worktree !== " " ? c.worktree : c.index || "");

/** Join the session's edited files with the working tree. Order is preserved (the Agent
 *  sends newest touch first); use sortRows to switch to path order.
 *
 *  `committed` is the repo-relative set from GET /sessions/{name}/committed — only
 *  consulted for rows the working tree has nothing to say about. */
export function joinChanges(files: SessionFile[], changes: FsChange[], committed?: string[]): FileRow[] {
  const wasCommitted = committed?.length ? new Set(committed) : null;
  const byKey = new Map<string, FsChange>();
  for (const c of changes) {
    const prefix = "repos/" + c.repo + "/";
    const rel = c.path.startsWith(prefix) ? c.path.slice(prefix.length) : c.path;
    byKey.set(changeKey(c.repo, rel), c);
  }
  return files.map((f) => {
    const rel = f.rel || "";
    const slash = rel.lastIndexOf("/");
    const change = f.repo && rel ? byKey.get(changeKey(f.repo, rel)) : undefined;
    let state: FileState = "clean";
    if (!f.repo || !rel) state = "outside";
    else if (change?.untracked) state = "untracked";
    else if (change) state = change.worktree && change.worktree !== " " ? "unstaged" : "staged";
    else if (wasCommitted?.has(rel)) state = "committed";
    return {
      ...f,
      name: (rel ? rel.slice(slash + 1) : f.path.split("/").pop()) || f.path,
      dir: slash > 0 ? rel.slice(0, slash) : "",
      state,
      deleted: f.verb === "delete" || (!!change && !change.untracked && codeOf(change) === "D"),
    };
  });
}

export function sortRows(rows: FileRow[], sort: FileSort): FileRow[] {
  if (sort === "recent") {
    // The Agent already sorts by newest touch; re-sorting locally keeps the toggle
    // reversible without another fetch.
    return [...rows].sort((a, b) => (b.lastTs || "").localeCompare(a.lastTs || "") || b.lastIdx - a.lastIdx);
  }
  return [...rows].sort(
    (a, b) => compareText(a.repo || "", b.repo || "") || compareText(a.rel || a.path, b.rel || b.path),
  );
}

/** The badge a row shows for its git state. `clean` and `outside` are muted on purpose:
 *  they are still part of the session's work, they just have no diff behind them. */
export function stateBadge(state: FileState): { cls: string; label: MsgKey } {
  switch (state) {
    case "untracked":
      return { cls: "st-add", label: "pj.st_untracked" };
    case "unstaged":
      return { cls: "st-mod", label: "mirror.files.st_unstaged" };
    case "staged":
      return { cls: "st-mod", label: "mirror.files.st_staged" };
    case "committed":
      return { cls: "st-done", label: "mirror.files.st_committed" };
    case "outside":
      return { cls: "st-muted", label: "mirror.files.st_outside" };
    default:
      return { cls: "st-muted", label: "mirror.files.st_clean" };
  }
}

/**
 * Open a row. The default gesture is the DIFF, because that is what "what did this
 * session do to this file" means — but only where git actually has one:
 *
 * an untracked file has no diff (`git diff` prints nothing), so opening one would hand
 * the reader an empty pane. Those rows, and rows with no working-tree change at all,
 * open the file instead — the same rule as the rail's Changes list.
 */
export function openRow(row: FileRow, split = false): void {
  if (row.repo && row.rel && (row.state === "unstaged" || row.state === "staged")) {
    if (split) {
      useLayoutStore
        .getState()
        .openTargetInNew(
          { content: { kind: "wtdiff", scmRepo: row.repo, filePath: row.rel, diffStaged: row.state === "staged" } },
          true,
        );
    } else {
      openFileDiff(row.repo, row.rel, row.state === "staged");
    }
    return;
  }
  if (row.deleted) return; // nothing left on disk to show
  if (split) {
    useLayoutStore.getState().openTargetInNew({ content: { kind: "file", filePath: row.path } }, true);
    return;
  }
  openFileMode(row.path, "view");
}

/**
 * Session → its edited files. The mirror's transcript poll is the writer (the list rides
 * that response, so there is no second poll and the strip can never show a different
 * moment than the turns above it); the command palette is a reader, and fetches once for
 * itself when the mirror for that session was never open.
 */
interface SessionFilesStore {
  bySession: Record<string, SessionFile[]>;
  set(session: string, files: SessionFile[]): void;
  forget(session: string): void;
}

export const useSessionFilesStore = create<SessionFilesStore>((set) => ({
  bySession: {},
  set: (session, files) => set((s) => ({ bySession: { ...s.bySession, [session]: files } })),
  forget: (session) =>
    set((s) => {
      const next = { ...s.bySession };
      delete next[session];
      return { bySession: next };
    }),
}));
